pyo3-macros-backend/src/method.rs

```diff
@@ -113,12 +113,14 @@ impl FnType {
             }
             FnType::FnClass | FnType::FnNewClass => {
                 quote! {
-                    _pyo3::types::PyType::from_type_ptr(py, _slf as *mut _pyo3::ffi::PyTypeObject),
+                    #[allow(clippy::useless_conversion)]
+                    ::std::convert::Into::into(_pyo3::types::PyType::from_type_ptr(py, _slf as *mut _pyo3::ffi::PyTypeObject)),
                 }
             }
             FnType::FnModule => {
                 quote! {
-                    py.from_borrowed_ptr::<_pyo3::types::PyModule>(_slf),
+                    #[allow(clippy::useless_conversion)]
+                    ::std::convert::Into::into(py.from_borrowed_ptr::<_pyo3::types::PyModule>(_slf)),
```

Comment:
No for a semantic question: `Into` is quite generic here. Do we want and/or have a specific trait to handle this? `IntoPy`?

Reply:
We need a trait that can return either `&PyModule` or `Py<PyModule>`, so `IntoPy` cannot do the job here. `Into` is the simplest choice, but it's true that it makes possible to use `PyObject` instead of `Py<PyType>`. (There is a macro-level check for `pass_module`, so it's not possible to use `PyObject,` but as I said, I would like to remove it to rely on compilation error)

I will add a new `PyOrRef` trait.

Reply:
> I will add a new ToPyOrRef trait.

Please wait until doing that to give others some time to chime in. Adding a new trait that becomes part of the `#[doc(hidden)]` macro API is a somewhat costly addition w.r.t. maintenance and I wouldn't want to rust it.

Reply:
Sorry, I have seen you comment after pushing the change. I'm rollbacking it for now.

Reply:
I would if we can press our usual traits for function argument extract into service here, i.e. `FromPyObject`?

Reply:
`FromPyObject` returns a `Result`, so it requires to add `.unwrap()`, but you would have the same issue as with `Into`, i.e. you can use `PyObject` instead of `Py<PyType>`. I still prefer using `Into` instead of `FromPyObject`.

Reply:
There's a strong overlap here with #3564, where we can't decide if we want to stay with `TryFrom`. I think `From` is better than `FromPyObject`, because `FromPyObject` will involve a needless type check (and is fallible as @wyfo observed).

I can't really see any better options than sticking with `From`/`TryFrom` and adding clippy allows, or adding traits to solve clippy complaints with our current `pass_module`, `classmethod` and method receiver implementations.

Given that the existing code works, and there's not really any harm with users taking the module as a `PyObject`, I'm leaning towards sticking with `From`/`TryFrom`.

Reply:
Let's stick to the appropriately infallible `From` then.